### PR TITLE
Validation dataset support

### PR DESCRIPTION
**Pull request by maxolek**
Validation-loss evaluation for `ValueTrainer`

**Change**

- Adds support for supplying an optional validation `DataReader` to `ValueTrainer::run`
- Adds a validation runner that evaluates the current model weights without modifying training state
- Adds forward loss evaluation support through `LossEvaluator`
- Uses the same WDL scheduler as training and maps validation targets using the current training `Step`
- Supports frequent validation or once per superbatch
    - Always evaluates at end of each superbatch, frequency of `0` disables additional validation
- Adds validation logging and records validation loss separately from training loss
- I’ve already integrated it into the examples

**Notes**
- User can choose to randomly sample N positions from the supplied validation set via `TestDataset::at(val_path).positions(N)`
   - If N is not specified, the entire dataset is loaded
- The entire dataset is loaded into memory once at training initiation
   - If randomly sampling N, only those N are kept in the cache during training, else the entire dataset is kept in memory
- The validation samples are kept the exact same during the entire training run for apples-apples comparison

**Caveats**

- I only tested this against stockfish binpacks (and it worked as expected), but I don't see why it would fail with others
- End-of-superbatch validation prints before training summary, but validation is still performed after training weights are updated/finalized
- I tried to match format as much as possible but I'm new here (and very commit happy so sorry about that) so I tried my best!

Validation is evaluated after the corresponding optimizer update, so end-of-superbatch validation reflects the fully trained state.

### CODE REVIEW

**Comment by jw1912** [comment]
Validation data was previously intentionally removed from the default bullet `ValueTrainer::run` method because it has been concluded to be a useless metric.

I am open to adding it back if there is enough demand, but not like this:
- Editing `ValueTrainer::run` is a breaking change for everyone
- `DataLoaderOnce` adds a lot of code, I would prefer reuse the existing dataloader code even though it would result in something functionally different
- The implementation as it stands is not performant
